README.md
# Node.js Back-End repository template based on Nix ecosystem

This template provides an environment that is fully reproducible on the other developer's machines and on our servers.

It allows everyone else to have a one-liner to install a set of tools and makes projects run on ALL Linux distributions and MacOS.

The project should serve as a starting point for Node.js backend services, listening to GraphQL and REST API requests. Everything has been already configured for you to get started with the development.

# TODO: Edit the following instructions to set up your project correctly.

- README.md:
  - Adjust title: **PROJECT-NAME Back-End repository with Apollo GraphQL**
  - Adjust description
  - Add information on project usage
- package.json:
  - Set proper name for project: _project-name-backend_
  - Adjust the general parts like description, repository url, author, scripts, etc.
- .env.example:
  - Add all necessary environment variables with relative value placeholder

**After completing all these steps, remove this section from the document**

## Table of contents

- [Development Guidelines](#development-guidelines)
- [Getting started](#getting-started)
- [Usage](#usage)
- [Project's Structure](#projects-structure)
- [Testing](#testing)
- [Release](#release)
- [Information](#information)
- [Support](#support)
- [Boilerplate](#boilerplate)

## Development guidelines

Before getting started with the development, make sure to check out the guide on our [development workflow](https://github.com/codeworks-projects/guide-development).

## Getting started

The following steps will get you everything you need to develop and make the project run on ALL Linux distributions and MacOS.

### 1. Make sure to have [Nix](https://nix.dev/tutorials/install-nix) installed, otherwise intall it with the following command:

```bash
sh <(curl -L https://nixos.org/nix/install)
```

### 2. [Setup direnv](https://nix.dev/tutorials/declarative-and-reproducible-developer-environments.html#direnv-automatically-activating-the-environment-on-directory-change) and automate nix shell reloading with the latest dependencies.

- Make sure to have [direnv](https://direnv.net/) installed.
- Hook it into your shell. For bash, add this line at the end of the `~/.bashrc` file: `eval "$(direnv hook bash)"`. Otherwise, visit https://direnv.net/docs/hook.html
- Run the following command on the top level of the project:

```bash
direnv allow
```

**Note:** it may take a while to download all tools and dependencies.

### 3. Run the following command and configure your local environemnt in the `.env` file:

```bash
cp .env.example .env
```

## Usage

### Configuration

If not else specified in the `.env` file, the _GraphQL_ and _REST API_ servers will start at http://localhost:4000/graphql and http://localhost:4001/api/v1, respectively.

Configure `.env` for database connection and setting of additional credentials and secrets.

### Development Tools

Everything is already set up for you, by following the 3-step [Getting Started](#getting-started) section.

If you need to add other development tools, **DO NOT** install them on your local machine! Follow to the [Nix](https://nix.dev) syntax to add new tools, by inserting them in the `buildInputs` list.

### That's it!

Run your application with:

- `yarn start` to build and start the application as in production.
- `yarn dev` to activate the Graphql playground and return mock data when it is missing for GraphQL queries.
- `yarn dev:mock` to activate the development mode and return mock data for the whole GraphQL schema. This is specially useful if you want to start the project without setting up a database.

_Tip: To use authentication and authorization on playground, enable `Include cookies` on Sandbox settings._

_Tip 2: To avoid filling up your terminal with useless logs, when using the playground, disable `Auto Update` in the Sandbox settings (Note: documentation on playground will not be refreshed on hot-reload)._

_Tip 3: GraphQL subscriptions are available on the playground, configuring the endpoint to `ws://localhost:4000/graphql` and the implementation to `graphql-ws`. You can test the susbcription `testNumberIncrement`, which should return sequential numbers every second. If it does not run, try out a different browser or contact the team._

Read also about the [project's structure](#projects-structure), [testing](#testing), and [release](#release).

## Project's Structure

This project is structured on 3 layers:

### API layer

The API layer includes GraphQL resolvers and handlers for REST endpoints, including also the necessary user authentication.

- `graphql/` includes middleware functions for logging and authentication, all the schema specification for types, inputs, queries and mutations, and all the resolvers. <br>
  Specification for the schema and implementation of resolvers should be divided into folders, based on the function they have on the whole system.

- `rest-api/` includes routers and functions for different types of REST APIs and webhooks.

### Service Layer

The service layer contains the business logic of the application, including more complex authorization operations.

Different services are divided into folders, based on the function they have on the whole system. This makes the whole system scalable, organized, and modular.

Furthermore, `services/` also includes a folder for the integrations with other external systems, e.g. email services, videocalling, payments, etc.

### Data Access Layer

Communication with the database happens in the data access layer.

Everything is managed by a singleton Database Connection Manager (_DBConnection_), which deals with the setup of the database schema, based on the specified models in the `data-access/models/` folder, and of a pool of connections on startup.

### Pub/Sub

The system also implements a pub/sub layer to handle actions that are not directly connected to the task performed by a function, e.g. notifying users, creating a change history, etc.

All subscribers are included in the `events/subscribers` folder.

To emit an event from one of the services (payment, in the example), use the following syntax:

```javascript
emitter.emit(events.PAYMENT_UPDATED, {
  service: "payment",
  data: {
    payment,
  },
});
```

### CronJob task handler

To execute repetive tasks, without the need of manual intervention, the system uses a cronjob task manager: `src/cron.ts`

## Testing

Testing is performed through the [Jest Testing Framework](https://jestjs.io/) and in the `src/__tests__/` folder. Simply, use one of the following commands to run tests that have been written:

- `yarn test`
- `yarn test:dev` to enable hot-reloading.

Mock integrations with external systems with `jest.spyOn().mockImplementation()`; multiple examples can be found in the existing test suites.

To test the backend, it is usually sufficient to write integration tests for the GraphQL layer, as a user would interact through a browser client.

### Coverage

To get an overview of what has been tested, run `yarn test:coverage` and then open [`coverage/lcov-report/index.html`](coverage/lcov-report/index.html) on the browser to see the results and what can be improved.
Coverage should be **above 85%** for all relevant functions.

## Release

Before creating a release:

- make sure that [testing](#testing) has been performed correctly (_Tip:_ run `LOG_LEVEL=fatal yarn test` to see the results, without the noise of logs);

- run `yarn test:coverage` to make sure all tests succeed and the minimum coverage of 85% is met;

- create a migration from the old database to the new one (Check out this documentation [here](https://sequelize.org/docs/v6/other-topics/migrations/);

- create a PR from `development` to `main` to visualize the changes that will be applied.

To create a release, from the `development` branch run:

```
$ yarn release
$ git push --follow-tags
```

#### Specific releases can be generated with:

- `yarn release -- --release-as minor` (increments the minor version)
- `yarn release -- --release-as patch` (increments the patch version)
- `yarn release -- --release-as alpha-1.1.0` (specific version)

This commands will take care of incrementing the version number, tagging the commit, and generating a changelog, based on the commit messages.

After a release has been created, it should be merged to the base branch with the following command from `main`:

```
git merge development --squash
```

The commit message should be the original `chore(release)` from development.

All these steps will ensure readability of the changes, simple but robust structure of the repository, and easy restoration of previous versions.

## Information

### About the project's ecosystem

This is a [Node.js](https://nodejs.org/en/) project, which is still a hybrid between JavaScript (for the graphql resolvers and api router functions) and Typescript, with [Apollo](https://www.apollographql.com/docs/apollo-server/) as _GraphQL_ framework, which enables interaction with the frontend. Instead, integration for external software should be available via _REST APIs_, using [Express](http://expressjs.com/).

_Note: If you have some spare time to upgrade also the remaining `.js` files to `.ts`, you are very welcome to do so (please, update the boilerplate as well). Just make sure that everything still runs properly after this operation :)_

For interaction with the database, the project uses [sequelize](https://sequelize.org/docs/v6/).

For logging information efficiently and with a low overhead, the project uses [pino](https://getpino.io/#/?id=low-overhead).

The ecosystem is based on [nix.dev](https://nix.dev) tutorials and the repository gets you started with [Nix](https://nixos.org/):

- [niv](https://github.com/nmattia/niv) for painless dependency management (aka pinning) with a weekly cronjob to bump dependencies
- [direnv](https://direnv.net/) for automatically loading your developer environment
- [gitignore.nix](https://github.com/hercules-ci/gitignore.nix) for respecting `.gitignore` when using your project as a source
- [pre-commit-hooks.nix](https://github.com/cachix/pre-commit-hooks.nix) for running linters (defaults to `shellcheck`, `nix-linter` and `nixpkgs-fmt`) when committing and on the CI
- [commitizen](https://github.com/commitizen/cz-cli) for writing commits in the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard. Having a convention on your commits makes it possible to parse them and use them for generating automatically the version or a [changelog](https://github.com/conventional-changelog/conventional-changelog)
- [standard-version](https://github.com/conventional-changelog/standard-version), utility for versioning using [Semantic Versioning](https://semver.org/) and CHANGELOG generation powered by Conventional Commits
- [GitHub Actions](https://github.com/features/actions) for CI with [dependabot](https://dependabot.com/) automatically bumping GitHub Actions versions

### Support

For support, please contact [hello@decandido.dev](mailto:hello@decandido.dev).

### Boilerplate

The project uses this boilerplate: https://github.com/lelebus/template-backend-nodejs


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 1.0.1 (2023-03-11)

## [2.0.0](https://github.com/codeworks-projects/template-nodejs-apollo/compare/v1.0.2...v2.0.0) (2022-12-10)


### Features

* activate mocks and improve other code parts ([98c39a0](https://github.com/codeworks-projects/template-nodejs-apollo/commit/98c39a08ad276dcd798d1941f00a3541d92d5b06))
* add DB_URL to .env.example ([42333b3](https://github.com/codeworks-projects/template-nodejs-apollo/commit/42333b3ceb1df7c5089de73f52c710bfe9d775bf))
* add logging + middleware ([c903f01](https://github.com/codeworks-projects/template-nodejs-apollo/commit/c903f01c395f0ffdec6c0d2fa8c6cece5545f993))
* add URL as GraphQL scalar type ([47f199c](https://github.com/codeworks-projects/template-nodejs-apollo/commit/47f199cee300e64e89432cfc5ac44c28410caef2))
* add users with auth middleware ([5e94d0b](https://github.com/codeworks-projects/template-nodejs-apollo/commit/5e94d0be78eae37f5d9dab21e4771bc0cb709f26))
* **api:** hello world example test ([d060966](https://github.com/codeworks-projects/template-nodejs-apollo/commit/d060966e1655835193994a841f2af50001f38351))
* clean example files ([cb6a556](https://github.com/codeworks-projects/template-nodejs-apollo/commit/cb6a5564e914546a4f62c613d2cb31cf577793ac))
* **db:** add virtual fields and other getters ([2f29ac1](https://github.com/codeworks-projects/template-nodejs-apollo/commit/2f29ac1d240f3fe5f4ae8430833b2cdb45ecfcf6))
* run example subscription ([a206f00](https://github.com/codeworks-projects/template-nodejs-apollo/commit/a206f00bdc9a9d496993f00c2c0cf0b59a6b7a61))
* upgrade ([1d8da72](https://github.com/codeworks-projects/template-nodejs-apollo/commit/1d8da72b7d822776e8c5174a4e28c50e6de737d1))
* upgrade to typescript ([9b7fdc0](https://github.com/codeworks-projects/template-nodejs-apollo/commit/9b7fdc017cdf4bf6c504f7d1dc5013f71b9836ce))
* **user:** setup tests ([671012c](https://github.com/codeworks-projects/template-nodejs-apollo/commit/671012c9a7ae5402101d325bb5e8c91993ac31cf))


### Bug Fixes

* **date:** serialization ([0647ca2](https://github.com/codeworks-projects/template-nodejs-apollo/commit/0647ca2d078dfcb106514e3e9a7a19846c627bd9))
* remove comment ([bafa8df](https://github.com/codeworks-projects/template-nodejs-apollo/commit/bafa8dfb744c561241f4b373cf9085a725a7e969))
* **sendgrid:** mock entirely on test mode ([01c3547](https://github.com/codeworks-projects/template-nodejs-apollo/commit/01c354793ad980c919892bd4642f90404c7aa6bf))

### [1.0.2](https://github.com/codeworks-projects/template-nodejs-apollo/compare/v1.0.1...v1.0.2) (2022-01-22)

### [1.0.1](https://github.com/codeworks-projects/template-nodejs-apollo/compare/v0.1.0...v1.0.1) (2022-01-18)

## [0.1.0](https://github.com/codeworks-projects/template-nodejs-apollo/compare/v0.0.2...v0.1.0) (2022-01-18)


### ⚠ BREAKING CHANGES

* Fully working template repo

### Features

* add commitizen ([aa87752](https://github.com/codeworks-projects/template-nodejs-apollo/commit/aa87752ad38568d69745aae667678c706f798c03))


* add link to development guide ([4721238](https://github.com/codeworks-projects/template-nodejs-apollo/commit/472123828052f5257e7476159b3ac0c89e0f4207))

### 0.0.2 (2022-01-17)


### Features

* add .env.example ([e8fa9ea](https://github.com/codeworks-projects/template-nodejs-apollo/commit/e8fa9eae1a79da13e8ec9d74a5a76fff52f8f4c3))
* add custom tools in shell.nix ([2f0750a](https://github.com/codeworks-projects/template-nodejs-apollo/commit/2f0750abfdec0f0486aaea02af043840f2246578))
* add dev dependencies ([9fc01d1](https://github.com/codeworks-projects/template-nodejs-apollo/commit/9fc01d1226d7da02fd297c1e3834da1afa62c434))
* add git pull as nix shell hook ([40b2c5f](https://github.com/codeworks-projects/template-nodejs-apollo/commit/40b2c5fdf49de9b0da8cd1170694cec8ca1c8756))
* basic working project ([b088354](https://github.com/codeworks-projects/template-nodejs-apollo/commit/b08835447481b187cc0421e895ccbce9bb9fff2b))
* make project run inside nix shell ([4bb6619](https://github.com/codeworks-projects/template-nodejs-apollo/commit/4bb6619b33b2c0967dc8d38897f478d31b0cf0bf))
* run yarn install as shell hook ([46b617a](https://github.com/codeworks-projects/template-nodejs-apollo/commit/46b617a31ed2e4150b4729aa6335a753706c40b6))
* separate graphql and rest api initialization ([980ee7e](https://github.com/codeworks-projects/template-nodejs-apollo/commit/980ee7e90b54b7156ab2995fcc9e8e525a80e101))
* update README ([0642efd](https://github.com/codeworks-projects/template-nodejs-apollo/commit/0642efd4af1d5e83d48047ec67df3b5842ada91b))


### Bug Fixes

* date graophql scalar resolver ([aeb69d8](https://github.com/codeworks-projects/template-nodejs-apollo/commit/aeb69d8684b186749d7142ee0043bf03b472b1b7))


src/config.ts
import dotenv from "dotenv";
dotenv.config();

export const jest = {
  jestWorkerId: process.env.JEST_WORKER_ID,
};

export const environment = process.env.NODE_ENV || "production";
export const log_level = process.env.LOG_LEVEL || "info";

export const graphql = {
  port: process.env.GRAPHQL_PORT || 4000,
  path: process.env.GRAPHQL_PATH || "/graphql",
  subscriptionTriggerNames: {
    NEW_NOTIFICATION: "NEW_NOTIFICATION",
  },
};

export const api = {
  port: process.env.API_PORT || 4001,
};

export const accessTokenSecret = process.env.ACCESS_TOKEN_SECRET;
export const accessTokenBody = {
  httpOnly: false,
  sameSite: "none",
  secure: true,
  maxAge: 900000, // 15 minutes
};

export const renewRefreshTokens = true;
export const refreshTokenSecret = process.env.REFRESH_TOKEN_SECRET;
export const refreshTokenBody = {
  httpOnly: true,
  sameSite: "none",
  secure: true,
  maxAge: 15 * 86400000, // 15 days
};

export const database = {
  url: process.env.DB_URL,
  name: process.env.DB_NAME,
  user: process.env.DB_USERNAME,
  password: process.env.DB_PASSWORD,
};

export const fallbackLocale = "IT";

export const sendgrid = {
  api_key: process.env.SENDGRID_API_KEY || "SG.undefined",
  verified_sender: {
    name: "Boilerplate",
    email: "info@example.com",
  },

  administrators: [
    {
      name: "Gabriele De Candido",
      email: "gabrieledecandido@flashbeing.com",
    },
    {
      name: "Alessio Crea",
      email: "alessiocrea@flashbeing.com",
    },
  ],

  templates: {
    SEND_PASSWORD_RESET_TOKEN: "d-undefined",
    SEND_EMAIL_CHANGE_TOKEN: "d-undefined",
  },
};

export const stripe = {
  secret_key: process.env.STRIPE_SECRET_KEY,
  webhook_secrets: {
    account: process.env.STRIPE_WHSEC_CHECKOUT_COMPLETED,
  },
};

export const aws_s3 = {
  accessKeyId: process.env.AWS_S3_ID,
  secretAccessKey: process.env.AWS_S3_SECRET,
  bucketName: process.env.AWS_S3_BUCKET_NAME || "undefined-bucket",
};

export const jitsi = {
  jitsiTokenPrivateKey: JSON.parse(`"${process.env.JITSI_TOKEN_PRIVATE_KEY}"`),
  jitsiAppId: process.env.JITSI_APP_ID,
  jitsiKid: process.env.JITSI_KID,
};
